File: backend/app/trust_models.py
from __future__ import annotations

import enum
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import CheckConstraint, DateTime, Enum, ForeignKey, Index, Integer, String, Text, UniqueConstraint, Uuid
from sqlalchemy.orm import Mapped, mapped_column

from .database import Base
from .security import utcnow


class VerificationMethod(str, enum.Enum):
    aadhaar_offline = "aadhaar_offline"
    government_id = "government_id"
    business = "business"


class VerificationStatus(str, enum.Enum):
    pending = "pending"
    verified = "verified"
    rejected = "rejected"
    expired = "expired"


class PaymentStatus(str, enum.Enum):
    created = "created"
    captured = "captured"
    failed = "failed"
    partially_refunded = "partially_refunded"
    refunded = "refunded"


class PayoutStatus(str, enum.Enum):
    pending = "pending"
    eligible = "eligible"
    processing = "processing"
    paid = "paid"
    held = "held"
    cancelled = "cancelled"


class RefundStatus(str, enum.Enum):
    pending = "pending"
    completed = "completed"
    failed = "failed"


class DisputeStatus(str, enum.Enum):
    open = "open"
    under_review = "under_review"
    resolved = "resolved"
    closed = "closed"


class RiskSignalStatus(str, enum.Enum):
    open = "open"
    reviewed = "reviewed"
    dismissed = "dismissed"


class IdentityVerification(Base):
    __tablename__ = "identity_verifications"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("users.id", ondelete="CASCADE"), index=True)
    method: Mapped[VerificationMethod] = mapped_column(Enum(VerificationMethod), nullable=False)
    status: Mapped[VerificationStatus] = mapped_column(Enum(VerificationStatus), default=VerificationStatus.pending, index=True)
    provider_name: Mapped[str] = mapped_column(String(80), default="external", nullable=False)
    provider_reference: Mapped[str] = mapped_column(String(160), unique=True, nullable=False)
    failure_reason: Mapped[str | None] = mapped_column(String(240))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)
    verified_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class Payment(Base):
    __tablename__ = "payments"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    booking_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("bookings.id", ondelete="CASCADE"), unique=True, index=True)
    customer_user_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("users.id"), index=True)
    provider_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("provider_profiles.id"), index=True)
    amount_paise: Mapped[int] = mapped_column(Integer, nullable=False)
    platform_fee_paise: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    refunded_paise: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="INR", nullable=False)
    gateway: Mapped[str] = mapped_column(String(60), nullable=False)
    gateway_reference: Mapped[str] = mapped_column(String(160), unique=True, nullable=False)
    status: Mapped[PaymentStatus] = mapped_column(Enum(PaymentStatus), default=PaymentStatus.created, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)
    captured_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    __table_args__ = (
        CheckConstraint("amount_paise >= 0", name="ck_payment_amount_nonnegative"),
        CheckConstraint("platform_fee_paise >= 0", name="ck_payment_fee_nonnegative"),
        CheckConstraint("refunded_paise >= 0", name="ck_payment_refund_nonnegative"),
    )


class ProviderPayout(Base):
    __tablename__ = "provider_payouts"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    payment_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("payments.id", ondelete="CASCADE"), unique=True, index=True)
    provider_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("provider_profiles.id"), index=True)
    amount_paise: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[PayoutStatus] = mapped_column(Enum(PayoutStatus), default=PayoutStatus.pending, index=True)
    hold_reason: Mapped[str | None] = mapped_column(String(240))
    gateway_reference: Mapped[str | None] = mapped_column(String(160), unique=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)
    eligible_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    __table_args__ = (CheckConstraint("amount_paise >= 0", name="ck_payout_amount_nonnegative"),)


class Dispute(Base):
    __tablename__ = "disputes"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    booking_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("bookings.id", ondelete="CASCADE"), index=True)
    opened_by_user_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("users.id"), index=True)
    against_user_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("users.id"), index=True)
    category: Mapped[str] = mapped_column(String(80), nullable=False)
    summary: Mapped[str] = mapped_column(Text, nullable=False)
    requested_refund_paise: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    status: Mapped[DisputeStatus] = mapped_column(Enum(DisputeStatus), default=DisputeStatus.open, index=True)
    resolution: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    __table_args__ = (CheckConstraint("requested_refund_paise >= 0", name="ck_dispute_refund_nonnegative"),)


class Refund(Base):
    __tablename__ = "refunds"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    payment_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("payments.id", ondelete="CASCADE"), index=True)
    dispute_id: Mapped[UUID | None] = mapped_column(Uuid, ForeignKey("disputes.id", ondelete="SET NULL"), index=True)
    amount_paise: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[RefundStatus] = mapped_column(Enum(RefundStatus), default=RefundStatus.pending, index=True)
    gateway_reference: Mapped[str | None] = mapped_column(String(160), unique=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    __table_args__ = (CheckConstraint("amount_paise > 0", name="ck_refund_amount_positive"),)


class Review(Base):
    __tablename__ = "reviews"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    booking_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("bookings.id", ondelete="CASCADE"), index=True)
    reviewer_user_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("users.id"), index=True)
    subject_user_id: Mapped[UUID] = mapped_column(Uuid, ForeignKey("users.id"), index=True)
    subject_provider_id: Mapped[UUID | None] = mapped_column(Uuid, ForeignKey("provider_profiles.id"), index=True)
    rating: Mapped[int] = mapped_column(Integer, nullable=False)
    comment: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)
    __table_args__ = (
        UniqueConstraint("booking_id", "reviewer_user_id", "subject_user_id", name="uq_booking_review_party"),
        CheckConstraint("rating >= 1 AND rating <= 5", name="ck_review_rating_range"),
    )


class RiskSignal(Base):
    __tablename__ = "risk_signals"
    id: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    entity_type: Mapped[str] = mapped_column(String(60), index=True)
    entity_id: Mapped[str] = mapped_column(String(80), index=True)
    kind: Mapped[str] = mapped_column(String(100), index=True)
    score: Mapped[int] = mapped_column(Integer, nullable=False)
    detail: Mapped[str | None] = mapped_column(Text)
    status: Mapped[RiskSignalStatus] = mapped_column(Enum(RiskSignalStatus), default=RiskSignalStatus.open, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    __table_args__ = (
        CheckConstraint("score >= 0 AND score <= 100", name="ck_risk_score_range"),
        Index("ix_risk_entity", "entity_type", "entity_id", "created_at"),
    )

File: lib/features/trust/domain/trust_models.dart
enum TrustVerificationMethod { aadhaarOffline, governmentId, business }

extension TrustVerificationMethodX on TrustVerificationMethod {
  String get wireName => switch (this) {
        TrustVerificationMethod.aadhaarOffline => 'aadhaar_offline',
        TrustVerificationMethod.governmentId => 'government_id',
        TrustVerificationMethod.business => 'business',
      };

  String get label => switch (this) {
        TrustVerificationMethod.aadhaarOffline => 'Aadhaar Offline e-KYC',
        TrustVerificationMethod.governmentId => 'Government ID',
        TrustVerificationMethod.business => 'Business verification',
      };
}

class TrustVerification {
  const TrustVerification({
    required this.id,
    required this.method,
    required this.status,
    required this.providerName,
    required this.providerReference,
    required this.createdAt,
    this.verifiedAt,
    this.failureReason,
  });

  final String id;
  final String method;
  final String status;
  final String providerName;
  final String providerReference;
  final DateTime createdAt;
  final DateTime? verifiedAt;
  final String? failureReason;

  factory TrustVerification.fromJson(Map<String, dynamic> json) =>
      TrustVerification(
        id: json['id'].toString(),
        method: json['method'] as String? ?? '',
        status: json['status'] as String? ?? 'pending',
        providerName: json['provider_name'] as String? ?? '',
        providerReference: json['provider_reference'] as String? ?? '',
        createdAt: _date(json['created_at']),
        verifiedAt: _nullableDate(json['verified_at']),
        failureReason: json['failure_reason'] as String?,
      );
}

class TrustPayment {
  const TrustPayment({
    required this.id,
    required this.bookingId,
    required this.amountPaise,
    required this.platformFeePaise,
    required this.refundedPaise,
    required this.currency,
    required this.gateway,
    required this.gatewayReference,
    required this.status,
    required this.createdAt,
    this.capturedAt,
  });

  final String id;
  final String bookingId;
  final int amountPaise;
  final int platformFeePaise;
  final int refundedPaise;
  final String currency;
  final String gateway;
  final String gatewayReference;
  final String status;
  final DateTime createdAt;
  final DateTime? capturedAt;

  double get amountRupees => amountPaise / 100;
  double get refundedRupees => refundedPaise / 100;

  factory TrustPayment.fromJson(Map<String, dynamic> json) => TrustPayment(
        id: json['id'].toString(),
        bookingId: json['booking_id'].toString(),
        amountPaise: json['amount_paise'] as int? ?? 0,
        platformFeePaise: json['platform_fee_paise'] as int? ?? 0,
        refundedPaise: json['refunded_paise'] as int? ?? 0,
        currency: json['currency'] as String? ?? 'INR',
        gateway: json['gateway'] as String? ?? '',
        gatewayReference: json['gateway_reference'] as String? ?? '',
        status: json['status'] as String? ?? 'created',
        createdAt: _date(json['created_at']),
        capturedAt: _nullableDate(json['captured_at']),
      );
}

class TrustPayout {
  const TrustPayout({
    required this.id,
    required this.paymentId,
    required this.amountPaise,
    required this.status,
    required this.createdAt,
    this.holdReason,
    this.eligibleAt,
    this.paidAt,
  });

  final String id;
  final String paymentId;
  final int amountPaise;
  final String status;
  final String? holdReason;
  final DateTime createdAt;
  final DateTime? eligibleAt;
  final DateTime? paidAt;

  double get amountRupees => amountPaise / 100;

  factory TrustPayout.fromJson(Map<String, dynamic> json) => TrustPayout(
        id: json['id'].toString(),
        paymentId: json['payment_id'].toString(),
        amountPaise: json['amount_paise'] as int? ?? 0,
        status: json['status'] as String? ?? 'pending',
        holdReason: json['hold_reason'] as String?,
        createdAt: _date(json['created_at']),
        eligibleAt: _nullableDate(json['eligible_at']),
        paidAt: _nullableDate(json['paid_at']),
      );
}

class BookingTrustState {
  const BookingTrustState({
    required this.canPay,
    required this.canStart,
    required this.canComplete,
    required this.canReview,
    required this.openDisputeCount,
    this.payment,
    this.payout,
  });

  final TrustPayment? payment;
  final TrustPayout? payout;
  final bool canPay;
  final bool canStart;
  final bool canComplete;
  final bool canReview;
  final int openDisputeCount;

  factory BookingTrustState.fromJson(Map<String, dynamic> json) =>
      BookingTrustState(
        payment: json['payment'] == null
            ? null
            : TrustPayment.fromJson(_map(json['payment'])),
        payout: json['payout'] == null
            ? null
            : TrustPayout.fromJson(_map(json['payout'])),
        canPay: json['can_pay'] as bool? ?? false,
        canStart: json['can_start'] as bool? ?? false,
        canComplete: json['can_complete'] as bool? ?? false,
        canReview: json['can_review'] as bool? ?? false,
        openDisputeCount: json['open_dispute_count'] as int? ?? 0,
      );
}

class TrustDispute {
  const TrustDispute({
    required this.id,
    required this.bookingId,
    required this.category,
    required this.summary,
    required this.requestedRefundPaise,
    required this.status,
    required this.createdAt,
    this.resolution,
    this.resolvedAt,
  });

  final String id;
  final String bookingId;
  final String category;
  final String summary;
  final int requestedRefundPaise;
  final String status;
  final String? resolution;
  final DateTime createdAt;
  final DateTime? resolvedAt;

  factory TrustDispute.fromJson(Map<String, dynamic> json) => TrustDispute(
        id: json['id'].toString(),
        bookingId: json['booking_id'].toString(),
        category: json['category'] as String? ?? '',
        summary: json['summary'] as String? ?? '',
        requestedRefundPaise: json['requested_refund_paise'] as int? ?? 0,
        status: json['status'] as String? ?? 'open',
        resolution: json['resolution'] as String?,
        createdAt: _date(json['created_at']),
        resolvedAt: _nullableDate(json['resolved_at']),
      );
}

class TrustOverview {
  const TrustOverview({
    required this.identityVerified,
    required this.paymentsCount,
    required this.payoutsCount,
    required this.openDisputesCount,
    this.latestVerification,
  });

  final bool identityVerified;
  final TrustVerification? latestVerification;
  final int paymentsCount;
  final int payoutsCount;
  final int openDisputesCount;

  factory TrustOverview.fromJson(Map<String, dynamic> json) => TrustOverview(
        identityVerified: json['identity_verified'] as bool? ?? false,
        latestVerification: json['latest_verification'] == null
            ? null
            : TrustVerification.fromJson(_map(json['latest_verification'])),
        paymentsCount: json['payments_count'] as int? ?? 0,
        payoutsCount: json['payouts_count'] as int? ?? 0,
        openDisputesCount: json['open_disputes_count'] as int? ?? 0,
      );
}

class ProviderReviewSummary {
  const ProviderReviewSummary({required this.count, this.averageRating});
  final int count;
  final double? averageRating;

  factory ProviderReviewSummary.fromJson(Map<String, dynamic> json) =>
      ProviderReviewSummary(
        count: json['count'] as int? ?? 0,
        averageRating: (json['average_rating'] as num?)?.toDouble(),
      );
}

Map<String, dynamic> _map(Object? value) {
  if (value is Map<String, dynamic>) return value;
  if (value is Map) return value.map((key, item) => MapEntry('$key', item));
  throw const FormatException('Expected JSON object');
}

DateTime _date(Object? value) =>
    DateTime.tryParse(value?.toString() ?? '')?.toLocal() ?? DateTime.now();

DateTime? _nullableDate(Object? value) =>
    value == null ? null : DateTime.tryParse(value.toString())?.toLocal();
